src/linear_algebra/bilinear_form.lean

```diff
@@ -1016,3 +1046,300 @@ begin
+  { exact submodule.mem_span_singleton.2 ⟨(B x y * (B x x)⁻¹), rfl⟩ },
+  { intros z hz,
+    rcases submodule.mem_span_singleton.1 hz with ⟨μ, rfl⟩,
+    simp [is_ortho, mul_assoc, inv_mul_cancel hx, hB x] },
+  { simp }
+end
+
+/-- Given a bilinear form `B` and some `x` such that `B x x ≠ 0`, the span of the singleton of `x`
+  is complement to its orthogonal complement. -/
+lemma is_compl_singleton {B : bilin_form K V}
+  (hB : sym_bilin_form.is_sym B) {x : V} (hx : ¬ B.is_ortho x x) :
+  is_compl (K ∙ x) (B.orthogonal $ K ∙ x) :=
+{ inf_le_bot := eq_bot_iff.1 $ span_inf_ortho_eq_bot hx,
+  top_le_sup := eq_top_iff.1 $ span_sup_ortho_eq_top hB hx }
+
+/-- The natural isomorphism between a singleton and the quotient by its orthogonal complement. -/
+noncomputable def quotient_equiv_of_ortho_singleton
+  {B : bilin_form K V} (hB : sym_bilin_form.is_sym B) {x : V} (hx : ¬ B.is_ortho x x) :=
+submodule.quotient_equiv_of_is_compl _ _ (is_compl_singleton hB hx)
+
+/-- The natural isomorphism from the product between a singleton and its orthogonal component
+  and the whole space. -/
+noncomputable def prod_equiv_of_ortho_singleton
+  {B : bilin_form K V} (hB : sym_bilin_form.is_sym B) {x : V} (hx : ¬ B.is_ortho x x) :=
+submodule.prod_equiv_of_is_compl _ _ (is_compl_singleton hB hx)
+
+lemma restrict_ortho_singleton_nondegenerate (B : bilin_form K V) (hB₁ : nondegenerate B)
+  (hB₂ : sym_bilin_form.is_sym B) {x : V} (hx : ¬ B.is_ortho x x) :
+  nondegenerate $ B.restrict $ B.orthogonal (K ∙ x) :=
+begin
+  refine λ m hm, submodule.coe_eq_zero.1 (hB₁ m.1 (λ n, _)),
+  have : n ∈ (K ∙ x) ⊔ B.orthogonal (K ∙ x) :=
+    (span_sup_ortho_eq_top hB₂ hx).symm ▸ submodule.mem_top,
+  rcases submodule.mem_sup.1 this with ⟨y, hy, z, hz, rfl⟩,
+  specialize hm ⟨z, hz⟩,
+  rw restrict at hm,
+  erw [add_right, show B m.1 y = 0, by exact m.2 y hy, hm, add_zero]
+end
+
+open finite_dimensional
+
+variable [finite_dimensional K V]
+
+lemma to_lin_surjective_of_finite_dimensional
+  {B : bilin_form K V} (hB : B.nondegenerate) :
+  function.surjective B.to_lin :=
+begin
+  classical,
+  refine (linear_map.injective_iff_surjective_of_findim_eq_findim
+    (linear_equiv.findim_eq _)).1 (to_lin_injective hB),
+  have hB := classical.some_spec (finite_dimensional.exists_is_basis_finite K V),
+  haveI := classical.choice hB.2,
+  exact is_basis.to_dual_equiv _ hB.1
+end
+
+lemma to_lin_range_eq_top_of_finite_dimensional
+  {B : bilin_form K V} (hB : B.nondegenerate) :
+  B.to_lin.range = ⊤ :=
+linear_map.range_eq_top.2 $ to_lin_surjective_of_finite_dimensional hB
+
+/-- Given a nondegenerate bilinear form `B`, `B.to_dual` is the isomorphism
+between a vector space and its dual with the underlying linear map `B.to_lin`. -/
+noncomputable def to_dual (B : bilin_form K V) (hB : B.nondegenerate) :
+  V ≃ₗ[K] module.dual K V :=
+linear_map.linear_equiv_of_ker_eq_bot B.to_lin
+  (linear_map.ker_eq_bot.2 $ to_lin_injective hB) subspace.findim_eq_dual
```

Comment:
```suggestion
  (nondegenerate_iff_ker_eq_bot.mp hB) subspace.findim_eq_dual
```
(Untested, but something like this -- deduce it from `nondegenerate_iff_ker_eq_bot` rather than converting and then unconverting to injectivity).

Then move this right after `nondegenerate_iff_ker_eq_bot`.  I would propose killing `to_lin_surjective_of_finite_dimensional`, `to_lin_range_eq_top_of_finite_dimensional`, and `to_lin_injective`, but if you really want them, deduce them from `to_dual`: use `linear_equiv.injective`, `linear_equiv.surjective`, `linear_equiv.range`.